tower-router/src/lib.rs

```diff
@@ -0,0 +1,398 @@
+/// # use futures_util::never::Never;
+/// # use tower_router::{OneshotService, ServiceGroup};
+/// type StdError = Box<dyn std::error::Error + Send + Sync + 'static>;
+/// struct MyService(u8);
+///
+/// impl Service<String> for MyService {
+///     type Response = ();
+///     type Error = StdError;
+///     type Future = Ready<Result<(), Self::Error>>;
+///
+///     fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+///         Poll::Ready(Ok(()))
+///     }
+///
+///     fn call(&mut self, req: String) -> Self::Future {
+///         println!("{}: {}", self.0, req);
+///         ready(Ok(()))
+///     }
+/// }
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let s = MyService(0);
+///     let s = OneshotService(s);
+///     let mut sg = ServiceGroup::new(vec![s]);
+///     let (tx, rx) = tokio::sync::oneshot::channel();
+///     sg.push(0, (tx, String::from("a")));
+///
+///     use futures_util::stream::StreamExt;
+///     let fut = sg.next().await.unwrap().unwrap();
+///     tokio::spawn(fut);
+///     rx.await;
+/// }
+/// ```
+#[derive(Debug)]
+pub struct OneshotService<S>(pub S);
+
+impl<S, Req, T> Service<(oneshot::Sender<T>, Req)> for OneshotService<S>
+where
+    S: Service<Req, Response = T, Error = StdError>,
+    S::Future: Send + 'static,
+    T: Send + 'static,
+{
+    type Response = ();
+    type Error = S::Error;
+    type Future = Pin<Box<dyn std::future::Future<Output = Result<(), S::Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.0.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: (oneshot::Sender<T>, Req)) -> Self::Future {
+        let (tx, req) = req;
+        let fut = self.0.call(req);
+        Box::pin(async move {
+            let res: T = fut.await?;
+            // unwrap_or_else -> panic to avoid constraining `Error` and `T` with Debug
+            tx.send(res)
+                .unwrap_or_else(|_| panic!("receiver hung up on Req result"));
+            Ok(())
+        })
+    }
+}
+
+/// `ServiceGroup` polls a group of services for them to become ready, buffering incoming requests in
+/// the meanwhile.
+///
+/// To avoid head-of-line blocking, `ServiceGroup` buffers an unbounded amount of requests to each
+/// `Service`.
+///
+/// As clients become ready, it returns futures (via `impl Stream`) which perform the request on
+/// the correct `Service`. It does *not* `.await` these `Future`s - this is left to the caller.
+/// This way, the caller is able to `.await` several returned `Future`s at once, instead of
+/// blocking on the completion of one at a time.
+///
+/// It will `poll_ready` only the services which have pending requests. Otherwise, we could
+/// end up repeatedly calling `poll_ready` on a service for which there are no requests.
+#[pin_project::pin_project]
+#[derive(Debug)]
+pub struct ServiceGroup<S, Req> {
+    cls: Vec<S>,
```

Comment:
This means each service must be the same type, correct? That is a pretty big limitation.

Reply:
For posterity: Lucio and I chatted about this elsewhere and decided that it probably makes sense to have two different "routers": one, like this one, which chooses among services that have the same request type, and one that chooses which service to forward to depending on the request type. This one should probably be called something like `tower-choose` or `tower-shard`, while the other is (perhaps) a more natural fit for the name `tower-router`.